### PR TITLE
Add declaration-block-no-duplicate-custom-properties

### LINKED ISSUE #5124: Add declaration-block-no-duplicate-custom-properties
> What is the problem you're trying to solve?

I think it would be better if this rule supports the ability to check CSS custom properties. For example we have color palette from designers with long complex naming. In this case, it is easy to make the mistake.

```css
:root {
  --companyBrandPaleRed: red;
  --companyBrandPaleRed: blue; // <- just wrong copy-paste
}
```

> What solution would you like to see?

I suggest a small optional addition to the existing rule options without breaking changes. 
```
'declaration-block-no-duplicate-properties': [true, {
	'lookAtCustomProperties': true,
}],
```

example PR: https://github.com/stylelint/stylelint/compare/master...voskresla:no-duplicate-properties-add-custom-properties

### PR DESCRIPTION
**Pull request by voskresla**
> Which issue, if any, is this issue related to?

Closes #5124

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

**Comment by voskresla** [comment]
is it ok with CodeQL errors?

File: docs/user-guide/rules/list.md [hunk 1]
```diff
@@ -141,6 +141,7 @@ Grouped first by the following categories and then by the [_thing_](http://apps.
 ### Custom property
 
 - [`custom-property-pattern`](../../../lib/rules/custom-property-pattern/README.md): Specify a pattern for custom properties.
+- [`declaration-block-no-duplicate-custom-properties`](../../../lib/rules/declaration-block-no-duplicate-custom-properties/README.md): Disallow duplicate custom properties within declaration blocks.
```
Comment (jeddy3) [inline docs/user-guide/rules/list.md:144]: Let's move this to the "Possible Errors -> Declaration Block" section (in alphabetical order).

I made a mistake in the blueprint, sorry.

File: lib/rules/declaration-block-no-duplicate-custom-properties/README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# declaration-block-no-duplicate-properties
```
Comment (jeddy3) [inline lib/rules/declaration-block-no-duplicate-custom-properties/README.md:1]:
```suggestion
# declaration-block-no-duplicate-custom-properties
```

File: lib/rules/declaration-block-no-duplicate-custom-properties/README.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+```
+
+<!-- prettier-ignore -->
+```css
+a { --custom-property: pink; --cUstOm-prOpErtY: orange; }
+```
+
+## Optional secondary options
+
+None
```
Comment (jeddy3) [inline lib/rules/declaration-block-no-duplicate-custom-properties/README.md:44]:
```suggestion
```

Let's remove as we don't normally have this title if there are no secondary options.

File: lib/rules/declaration-block-no-duplicate-custom-properties/__tests__/index.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+			code: 'a { --custom-property: 1 }',
+		},
+		{
+			code: 'a { --custom-property: 1; --cUstOm-prOpErtY: 1 }',
+		},
+		{
+			code: 'a { --custom-property: 1; color: pink; --cUstOm-prOpErtY: 1 }',
+		},
+		{
+			code: 'a { color: var(--custom-property, --custom-property) }',
```
Comment (jeddy3) [inline lib/rules/declaration-block-no-duplicate-custom-properties/__tests__/index.js:20]: 👍

**Review by jeddy3 — CHANGES_REQUESTED** [review]
@voskresla Thanks, looking good!

I've requested some minor documentation changes.

> is it ok with CodeQL errors?

Yes. They're unrelated to this PR.

File: lib/rules/declaration-block-no-duplicate-custom-properties/index.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+function rule(on) {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, { actual: on });
+
+		if (!validOptions) {
+			return;
+		}
+
+		eachDeclarationBlock(root, (eachDecl) => {
+			const decls = [];
```
Comment (vankop) [inline lib/rules/declaration-block-no-duplicate-custom-properties/index.js:27]: better use `Set` for performance

File: lib/rules/declaration-block-no-duplicate-custom-properties/index.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+				const indexDuplicate = decls.indexOf(prop);
+
+				if (indexDuplicate !== -1) {
+					report({
+						message: messages.rejected(prop),
+						node: decl,
+						result,
+						ruleName,
+					});
```
Comment (vankop) [inline lib/rules/declaration-block-no-duplicate-custom-properties/index.js:48]:
```suggestion
					});return;
```
don't need to add custom property again on line 51

**Review by vankop — APPROVED** [review]

**Review by jeddy3 — APPROVED** [review]
@voskresla Nicely done! LGTM.

**Comment by voskresla** [comment]
big thx @vankop @jeddy3 for the perfect support.

**Comment by jeddy3** [comment]
Changelog entry added:

> - Added: `declaration-block-no-duplicate-custom-properties` rule ([#5125](https://github.com/stylelint/stylelint/pull/5125)).

**Pull request merged**